### PR TITLE
Adds automatic mypy typing and cleans up a bunch of typing errors

### PR DESCRIPTION
**Pull request by evanhanders**
One of my biggest struggles as an outsider trying to use this code has been trying to figure out what function inputs/outputs are. There's a lot of levels of inheritance and often that means opening a few files to get to the bottom of what each variable is.

So -- this PR beefs up the typing in the repository! I've added a .github/workflows/mypy.yml file that automatically uses mypy to type-check the code for python versions 3.10-3.12. With that, I've also added typing calls to all function definitions.

At the time of this PR, all tests in the test/ folder are passing for me.

Note that this PR builds on the minor repo reorg from PR #15 .

### CODE REVIEW

File: iit/tasks/ioi/test_ioi.py [hunk 1]
```diff
@@ -154,16 +157,16 @@ def test_all_nodes_patching():
...
     # Not implemented yet
-    test_names = t.tensor(range(10, 60, 1))
+    test_names = Tensor(range(10, 60, 1))
     hl_model = IOI_HL(d_vocab=61, names=test_names)
 
-    aba = t.tensor([[1, 2, -1, 3, 4, -2, 5, 6, -1]])
-    abb = t.tensor([[1, 2, -1, 3, 4, -2, 5, 6, -2]])
-    baa = t.tensor([[1, 2, -2, 3, 4, -1, 5, 6, -1]])
-    bab = t.tensor([[1, 2, -2, 3, 4, -1, 5, 6, -2]])
+    aba = Tensor([[1, 2, -1, 3, 4, -2, 5, 6, -1]])
```
Comment (rhaps0dy) [inline iit/tasks/ioi/test_ioi.py]: I believe you're supposed to construct this with `t.tensor` like it was.

Comment (rhaps0dy) [reply iit/tasks/ioi/test_ioi.py]: https://pytorch.org/docs/stable/tensors.html#tensor-class-reference

Comment (evanhanders) [reply iit/tasks/ioi/test_ioi.py]: Oh, thanks! I'll put back these changes in a commit shortly.

Comment (evanhanders) [reply iit/tasks/ioi/test_ioi.py]: Fixed in https://github.com/cybershiptrooper/iit/pull/16/commits/b3834105008d7efe74d6e8fac7c232297ae6ea7c

**Comment by evanhanders** [comment]
One thing that should be discussed (and that I can make changes to): how do y'all want Tensors and torch to be referred to throughout the module?

Some files `import torch` then do `torch.Tensor`. Some files `import torch as t` then to `t.Tensor`. Some files import torch and separately `from torch import Tensor` then use that. I think the first few files I saw do:

```
import torch as t
from torch import Tensor
```

Then use Tensor for typing and t for all the other pytorch calls. I tried to make that standard, but could go through and replace with a different standard. But -- this is why there's a lot of changes of this form.

**Comment by rhaps0dy** [comment]
I don't think we need to standardize as this -- whatever is already there can stay.

I would lean towards using `t.Tensor` when it's already available, and otherwise using `from torch import Tensor`.

**Comment by rhaps0dy** [comment]
Can we merge this? @cybershiptrooper

**Comment by cybershiptrooper** [comment]
@evanhanders @rhaps0dy 

Yikes! Didn't see these PRs at all- apparently I don't get mailed anymore about these. Will fix this. Am adding everybody as a contributor as well.

**Pull request merged**